`esp-println` fails to compile with `defmt-espflash,auto` features

At the moment (commit `65b956d6`), logging with defmt is broken.

Steps:

```
$ cd esp-println

$ cargo build --features=esp32c3,log,defmt-espflash --target=riscv32imc-unknown-none-elf
   Compiling proc-macro2 v1.0.85
   Compiling version_check v0.9.4
   Compiling thiserror v1.0.61
   Compiling proc-macro-error-attr v1.0.4
   Compiling syn v1.0.109
   Compiling quote v1.0.36
   Compiling syn v2.0.66
   Compiling proc-macro-error v1.0.4
   Compiling defmt-macros v0.3.9
   Compiling defmt v0.3.8
   Compiling bitflags v1.3.2
   Compiling thiserror-impl v1.0.61
   Compiling esp-build v0.1.0 (/home/ubuntu/esp-hal/esp-build)
   Compiling esp-println v0.9.1 (/home/ubuntu/esp-hal/esp-println)
   Compiling defmt-parser v0.3.4
error[E0599]: no method named `flush` found for struct `Printer` in the current scope
   --> src/defmt.rs:55:17
    |
55  |         Printer.flush();
    |                 ^^^^^ this is an associated function, not a method
    |
   ::: src/lib.rs:83:1
    |
83  | pub struct Printer;
    | ------------------ method `flush` not found for this struct
    |
    = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
note: candidate #1 is defined in an impl for the type `Printer`
   --> src/lib.rs:170:5
    |
170 |     pub fn flush() {
    |     ^^^^^^^^^^^^^^
note: candidate #2 is defined in the trait `defmt::Logger`
   --> /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/defmt-0.3.8/src/traits.rs:105:5
    |
105 |     unsafe fn flush();
    |     ^^^^^^^^^^^^^^^^^^
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `flush`, perhaps you need to implement one of them:
            candidate #1: `Functions`
            candidate #2: `Log`
help: disambiguate the associated function for candidate #2
    |
55  |         <Printer as defmt::Logger>::flush();
    |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0599]: no method named `flush` found for struct `Printer` in the current scope
   --> src/defmt.rs:77:17
    |
77  |         Printer.flush();
    |                 ^^^^^ this is an associated function, not a method
    |
   ::: src/lib.rs:83:1
    |
83  | pub struct Printer;
    | ------------------ method `flush` not found for this struct
    |
    = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
note: candidate #1 is defined in an impl for the type `Printer`
   --> src/lib.rs:170:5
    |
170 |     pub fn flush() {
    |     ^^^^^^^^^^^^^^
note: candidate #2 is defined in the trait `defmt::Logger`
   --> /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/defmt-0.3.8/src/traits.rs:105:5
    |
105 |     unsafe fn flush();
    |     ^^^^^^^^^^^^^^^^^^
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `flush`, perhaps you need to implement one of them:
            candidate #1: `Functions`
            candidate #2: `Log`
help: disambiguate the associated function for candidate #2
    |
77  |         <Printer as defmt::Logger>::flush();
    |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0599]: no method named `write_bytes_assume_cs` found for struct `Printer` in the current scope
   --> src/defmt.rs:88:13
    |
88  |     Printer.write_bytes_assume_cs(bytes)
    |     --------^^^^^^^^^^^^^^^^^^^^^-------
    |     |       |
    |     |       this is an associated function, not a method
    |     help: use associated function syntax instead: `Printer::write_bytes_assume_cs(bytes)`
    |
   ::: src/lib.rs:83:1
    |
83  | pub struct Printer;
    | ------------------ method `write_bytes_assume_cs` not found for this struct
    |
    = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
note: the candidate is defined in an impl for the type `Printer`
   --> src/lib.rs:157:5
    |
157 |     pub fn write_bytes_assume_cs(bytes: &[u8]) {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0599`.
error: could not compile `esp-println` (lib) due to 3 previous errors
```

Work-around:

- If you need to use `defmt`, define either `uart` or `jtag-serial` explicitly. The issue is with the combination of `auto` and `defmt-espflash`

Well, let's try to replicate this first. 

The way I got to it was running, in the top level:

```
$ cargo xtask run-example esp-hal esp32c6 adc_cal
```

Gives:

```
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 1.94s
     Running `target/debug/xtask run-example esp-hal esp32c6 adc_cal`
[2024-06-15T07:32:51Z WARN  xtask] Package 'esp-hal' specified, using 'examples' instead
[2024-06-15T07:32:52Z INFO  xtask] Building example '/home/ubuntu/esp-hal/examples/src/bin/adc_cal.rs' for 'esp32c6'
[2024-06-15T07:32:52Z INFO  xtask] Package: "src/bin/adc_cal.rs"
   Compiling esp-println v0.9.1 (/home/ubuntu/esp-hal/esp-println)
   Compiling examples v0.0.0 (/home/ubuntu/esp-hal/examples)
error[E0599]: no method named `flush` found for struct `Printer` in the current scope
   --> /home/ubuntu/esp-hal/esp-println/src/defmt.rs:55:17
    |
55  |         Printer.flush();
    |         --------^^^^^--
    |         |       |
    |         |       this is an associated function, not a method
    |         help: use associated function syntax instead: `Printer::flush()`
    |
   ::: /home/ubuntu/esp-hal/esp-println/src/lib.rs:83:1
    |
83  | pub struct Printer;
    | ------------------ method `flush` not found for this struct
    |
    = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
note: the candidate is defined in the trait `defmt::Logger`
   --> /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/defmt-0.3.8/src/traits.rs:105:5
    |
105 |     unsafe fn flush();
    |     ^^^^^^^^^^^^^^^^^^
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `flush`, perhaps you need to implement one of them:
            candidate #1: `Functions`
            candidate #2: `Log`

error[E0599]: no method named `flush` found for struct `Printer` in the current scope
   --> /home/ubuntu/esp-hal/esp-println/src/defmt.rs:77:17
    |
77  |         Printer.flush();
    |         --------^^^^^--
    |         |       |
    |         |       this is an associated function, not a method
    |         help: use associated function syntax instead: `Printer::flush()`
    |
   ::: /home/ubuntu/esp-hal/esp-println/src/lib.rs:83:1
    |
83  | pub struct Printer;
    | ------------------ method `flush` not found for this struct
    |
    = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
note: the candidate is defined in the trait `defmt::Logger`
   --> /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/defmt-0.3.8/src/traits.rs:105:5
    |
105 |     unsafe fn flush();
    |     ^^^^^^^^^^^^^^^^^^
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `flush`, perhaps you need to implement one of them:
            candidate #1: `Functions`
            candidate #2: `Log`

error[E0599]: no method named `write_bytes_assume_cs` found for struct `Printer` in the current scope
  --> /home/ubuntu/esp-hal/esp-println/src/defmt.rs:88:13
   |
88 |     Printer.write_bytes_assume_cs(bytes)
   |             ^^^^^^^^^^^^^^^^^^^^^ method not found in `Printer`
   |
  ::: /home/ubuntu/esp-hal/esp-println/src/lib.rs:83:1
   |
83 | pub struct Printer;
   | ------------------ method `write_bytes_assume_cs` not found for this struct

For more information about this error, try `rustc --explain E0599`.
error: could not compile `esp-println` (lib) due to 3 previous errors
Error: Failed to execute cargo subcommand
```

The output in my original post is different. I must have had some local modifications in that code. Sorry 'bout that.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -207,7 +207,15 @@ jobs:
             --features=${{ matrix.device.soc }},log \
             --target=${{ matrix.device.target }} \
             esp-println
-        
+
+      # So #1678 doesn't reoccur ('defmt-espflash,auto')
+      - name: Build (with feature 'defmt-espflash')
+        run: |
+          cargo xtask build-package \
+            --features=${{ matrix.device.soc }},log,defmt-espflash \
+            --target=${{ matrix.device.target }} \
+            esp-println
+
   extras:
     runs-on: ubuntu-latest
 

--- a/esp-println/build.rs
+++ b/esp-println/build.rs
@@ -10,8 +10,7 @@ fn main() {
     assert_unique_used_features!("jtag-serial", "uart", "auto");
 
     // Ensure that, if the `jtag-serial` communication method feature is enabled,
-    // either the `esp32c3`, `esp32c6`, `esp32h2`, or `esp32s3` chip feature is
-    // enabled.
+    // a compatible chip feature is also enabled.
     if cfg!(feature = "jtag-serial")
         && !(cfg!(feature = "esp32c3")
             || cfg!(feature = "esp32c6")

--- a/esp-println/src/defmt.rs
+++ b/esp-println/src/defmt.rs
@@ -4,7 +4,7 @@
 #[cfg(feature = "critical-section")]
 use critical_section::RestoreState;
 
-use super::Printer;
+use super::PrinterImpl;
 
 /// Global logger lock.
 #[cfg(feature = "critical-section")]
@@ -52,7 +52,7 @@ unsafe impl defmt::Logger for Logger {
         // section.
         ENCODER.end_frame(do_write);
 
-        Printer.flush();
+        Self::flush();
 
         #[cfg(feature = "critical-section")]
         {
@@ -74,7 +74,7 @@ unsafe impl defmt::Logger for Logger {
     }
 
     unsafe fn flush() {
-        Printer.flush();
+        PrinterImpl::flush();
     }
 
     unsafe fn write(bytes: &[u8]) {
@@ -85,5 +85,5 @@ unsafe impl defmt::Logger for Logger {
 }
 
 fn do_write(bytes: &[u8]) {
-    Printer.write_bytes_assume_cs(bytes)
+    PrinterImpl::write_bytes_assume_cs(bytes)
 }

--- a/esp-println/src/lib.rs
+++ b/esp-println/src/lib.rs
@@ -68,18 +68,6 @@ macro_rules! dbg {
     };
 }
 
-#[cfg(any(feature = "jtag-serial", feature = "auto"))]
-pub struct PrinterSerialJtag;
-
-#[cfg(any(feature = "uart", feature = "auto"))]
-pub struct PrinterUart;
-
-#[cfg(feature = "jtag-serial")]
-pub type PrinterImpl = PrinterSerialJtag;
-
-#[cfg(feature = "uart")]
-pub type PrinterImpl = PrinterUart;
-
 pub struct Printer;
 
 impl core::fmt::Write for Printer {
@@ -89,58 +77,64 @@ impl core::fmt::Write for Printer {
     }
 }
 
-#[cfg(feature = "auto")]
-pub struct PrinterAuto;
+impl Printer {
+    fn write_bytes(bytes: &[u8]) {
+        with(|| {
+            PrinterImpl::write_bytes_assume_cs(bytes);
+            PrinterImpl::flush();
+        })
+    }
+}
+
+#[cfg(feature = "jtag-serial")]
+type PrinterImpl = serial_jtag_printer::Printer;
+
+#[cfg(feature = "uart")]
+type PrinterImpl = uart_printer::Printer;
 
 #[cfg(feature = "auto")]
-type PrinterImpl = PrinterAuto;
+type PrinterImpl = auto_printer::Printer;
 
-#[cfg(feature = "auto")]
+#[cfg(all(
+    feature = "auto",
+    any(
+        feature = "esp32c3",
+        feature = "esp32c6",
+        feature = "esp32h2",
+        feature = "esp32p4",    // as mentioned in 'build.rs'
+        feature = "esp32s3"
+    )
+))]
 mod auto_printer {
-    use super::{with, PrinterSerialJtag, PrinterUart};
+    use super::with;
+    use crate::{
+        serial_jtag_printer::Printer as PrinterSerialJtag,
+        uart_printer::Printer as PrinterUart,
+    };
 
-    impl super::PrinterAuto {
+    pub struct Printer;
+    impl Printer {
         fn use_jtag() -> bool {
-            #[cfg(any(
-                feature = "esp32c3",
-                feature = "esp32c6",
-                feature = "esp32h2",
-                feature = "esp32p4",    // as mentioned in 'build.rs'
-                feature = "esp32s3"
-            ))]
-            {
-                // Decide if serial-jtag is used by checking SOF interrupt flag.
-                // SOF packet is sent by the HOST every 1ms on a full speed bus.
-                // Between two consecutive ticks, there will be at least 1ms (selectable tick
-                // rate range is 1 - 1000Hz).
-                // We don't reset the flag - if it was ever connected we assume serial-jtag is
-                // used
-                #[cfg(feature = "esp32c3")]
-                const USB_DEVICE_INT_RAW: *const u32 = 0x60043008 as *const u32;
-                #[cfg(feature = "esp32c6")]
-                const USB_DEVICE_INT_RAW: *const u32 = 0x6000f008 as *const u32;
-                #[cfg(feature = "esp32h2")]
-                const USB_DEVICE_INT_RAW: *const u32 = 0x6000f008 as *const u32;
-                #[cfg(feature = "esp32p4")]
-                const USB_DEVICE_INT_RAW: *const u32 = unimplemented!();
-                #[cfg(feature = "esp32s3")]
-                const USB_DEVICE_INT_RAW: *const u32 = 0x60038000 as *const u32;
+            // Decide if serial-jtag is used by checking SOF interrupt flag.
+            // SOF packet is sent by the HOST every 1ms on a full speed bus.
+            // Between two consecutive ticks, there will be at least 1ms (selectable tick
+            // rate range is 1 - 1000Hz).
+            // We don't reset the flag - if it was ever connected we assume serial-jtag is
+            // used
+            #[cfg(feature = "esp32c3")]
+            const USB_DEVICE_INT_RAW: *const u32 = 0x60043008 as *const u32;
+            #[cfg(feature = "esp32c6")]
+            const USB_DEVICE_INT_RAW: *const u32 = 0x6000f008 as *const u32;
+            #[cfg(feature = "esp32h2")]
+            const USB_DEVICE_INT_RAW: *const u32 = 0x6000f008 as *const u32;
+            #[cfg(feature = "esp32p4")]
+            const USB_DEVICE_INT_RAW: *const u32 = unimplemented!();
+            #[cfg(feature = "esp32s3")]
+            const USB_DEVICE_INT_RAW: *const u32 = 0x60038000 as *const u32;
 
-                const SOF_INT_MASK: u32 = 0b10;
+            const SOF_INT_MASK: u32 = 0b10;
 
-                unsafe { (USB_DEVICE_INT_RAW.read_volatile() & SOF_INT_MASK) != 0 }
-            }
-
-            #[cfg(not(any(
-                feature = "esp32c3",
-                feature = "esp32c6",
-                feature = "esp32h2",
-                feature = "esp32p4",
-                feature = "esp32s3"
-            )))]
-            {
-                false
-            }
+            unsafe { (USB_DEVICE_INT_RAW.read_volatile() & SOF_INT_MASK) != 0 }
         }
 
         pub fn write_bytes_assume_cs(bytes: &[u8]) {
@@ -169,13 +163,19 @@ mod auto_printer {
     }
 }
 
-impl Printer {
-    pub fn write_bytes(bytes: &[u8]) {
-        with(|| {
-            PrinterImpl::write_bytes_assume_cs(bytes);
-            PrinterImpl::flush();
-        })
-    }
+#[cfg(all(
+    feature = "auto",
+    not(any(
+        feature = "esp32c3",
+        feature = "esp32c6",
+        feature = "esp32h2",
+        feature = "esp32p4",
+        feature = "esp32s3"
+    ))
+))]
+mod auto_printer {
+    // models that only have UART
+    pub type Printer = crate::uart_printer::Printer;
 }
 
 #[cfg(all(
@@ -190,6 +190,7 @@ impl Printer {
 ))]
 mod serial_jtag_printer {
     use portable_atomic::{AtomicBool, Ordering};
+    pub struct Printer;
 
     #[cfg(feature = "esp32c3")]
     const SERIAL_JTAG_FIFO_REG: usize = 0x6004_3000;
@@ -246,7 +247,7 @@ mod serial_jtag_printer {
         true
     }
 
-    impl super::PrinterSerialJtag {
+    impl Printer {
         pub fn write_bytes_assume_cs(bytes: &[u8]) {
             if fifo_full() {
                 // The FIFO is full. Let's see if we can progress.
@@ -289,7 +290,9 @@ mod serial_jtag_printer {
 #[cfg(all(any(feature = "uart", feature = "auto"), feature = "esp32"))]
 mod uart_printer {
     const UART_TX_ONE_CHAR: usize = 0x4000_9200;
-    impl super::PrinterUart {
+
+    pub struct Printer;
+    impl Printer {
         pub fn write_bytes_assume_cs(bytes: &[u8]) {
             for &b in bytes {
                 unsafe {
@@ -306,7 +309,8 @@ mod uart_printer {
 
 #[cfg(all(any(feature = "uart", feature = "auto"), feature = "esp32s2"))]
 mod uart_printer {
-    impl super::PrinterUart {
+    pub struct Printer;
+    impl Printer {
         pub fn write_bytes_assume_cs(bytes: &[u8]) {
             // On ESP32-S2 the UART_TX_ONE_CHAR ROM-function seems to have some issues.
             for chunk in bytes.chunks(64) {
@@ -445,7 +449,8 @@ mod uart_printer {
         }
     }
 
-    impl super::PrinterUart {
+    pub struct Printer;
+    impl Printer {
         pub fn write_bytes_assume_cs(bytes: &[u8]) {
             for chunk in bytes.chunks(Device::CHUNK_SIZE) {
                 for &b in chunk {

--- a/esp-println/src/lib.rs
+++ b/esp-println/src/lib.rs
@@ -89,67 +89,90 @@ impl core::fmt::Write for Printer {
     }
 }
 
-impl Printer {
-    #[cfg(not(feature = "auto"))]
-    pub fn write_bytes(bytes: &[u8]) {
-        with(|| {
-            PrinterImpl::write_bytes_assume_cs(bytes);
-            PrinterImpl::flush();
-        })
-    }
+#[cfg(feature="auto")]
+pub struct PrinterAuto;
 
-    #[cfg(feature = "auto")]
-    pub fn write_bytes(bytes: &[u8]) {
-        #[cfg(any(
-            feature = "esp32c3",
-            feature = "esp32c6",
-            feature = "esp32h2",
-            feature = "esp32s3"
-        ))]
-        {
-            // Decide if serial-jtag is used by checking SOF interrupt flag.
-            // SOF packet is sent by the HOST every 1ms on a full speed bus.
-            // Between two consecutive ticks, there will be at least 1ms (selectable tick
-            // rate range is 1 - 1000Hz).
-            // We don't reset the flag - if it was ever connected we assume serial-jtag is
-            // used
+#[cfg(feature = "auto")]
+type PrinterImpl = PrinterAuto;
 
-            #[cfg(feature = "esp32c3")]
-            const USB_DEVICE_INT_RAW: *const u32 = 0x60043008 as *const u32;
-            #[cfg(feature = "esp32c6")]
-            const USB_DEVICE_INT_RAW: *const u32 = 0x6000f008 as *const u32;
-            #[cfg(feature = "esp32h2")]
-            const USB_DEVICE_INT_RAW: *const u32 = 0x6000f008 as *const u32;
-            #[cfg(feature = "esp32s3")]
-            const USB_DEVICE_INT_RAW: *const u32 = 0x60038000 as *const u32;
+#[cfg(feature="auto")]
+mod auto_printer {
+    use super::{ with, PrinterSerialJtag, PrinterUart };
 
-            const SOF_INT_MASK: u32 = 0b10;
+    impl super::PrinterAuto {
+        fn use_jtag() -> bool {
+            #[cfg(any(
+                feature = "esp32c3",
+                feature = "esp32c6",
+                feature = "esp32h2",
+                feature = "esp32p4",    // as mentioned in 'build.rs'
+                feature = "esp32s3"
+            ))]
+            {
+                // Decide if serial-jtag is used by checking SOF interrupt flag.
+                // SOF packet is sent by the HOST every 1ms on a full speed bus.
+                // Between two consecutive ticks, there will be at least 1ms (selectable tick
+                // rate range is 1 - 1000Hz).
+                // We don't reset the flag - if it was ever connected we assume serial-jtag is
+                // used
+                #[cfg(feature = "esp32c3")]
+                const USB_DEVICE_INT_RAW: *const u32 = 0x60043008 as *const u32;
+                #[cfg(feature = "esp32c6")]
+                const USB_DEVICE_INT_RAW: *const u32 = 0x6000f008 as *const u32;
+                #[cfg(feature = "esp32h2")]
+                const USB_DEVICE_INT_RAW: *const u32 = 0x6000f008 as *const u32;
+                #[cfg(feature = "esp32p4")]
+                const USB_DEVICE_INT_RAW: *const u32 = unimplemented!();
+                #[cfg(feature = "esp32s3")]
+                const USB_DEVICE_INT_RAW: *const u32 = 0x60038000 as *const u32;
 
-            let is_serial_jtag =
-                unsafe { (USB_DEVICE_INT_RAW.read_volatile() & SOF_INT_MASK) != 0 };
+                const SOF_INT_MASK: u32 = 0b10;
 
-            if is_serial_jtag {
+                unsafe { (USB_DEVICE_INT_RAW.read_volatile() & SOF_INT_MASK) != 0 }
+            }
+
+            #[cfg(not(any(
+                feature = "esp32c3",
+                feature = "esp32c6",
+                feature = "esp32h2",
+                feature = "esp32p4",
+                feature = "esp32s3"
+            )))]
+            { false }
+        }
+
+        pub fn write_bytes_assume_cs(bytes: &[u8]) {
+            if Self::use_jtag() {
                 with(|| {
                     PrinterSerialJtag::write_bytes_assume_cs(bytes);
-                    PrinterSerialJtag::flush();
                 })
             } else {
                 with(|| {
                     PrinterUart::write_bytes_assume_cs(bytes);
+                })
+            }
+        }
+
+        pub fn flush() {
+            if Self::use_jtag() {
+                with(|| {
+                    PrinterSerialJtag::flush();
+                })
+            } else {
+                with(|| {
                     PrinterUart::flush();
                 })
             }
         }
 
-        #[cfg(not(any(
-            feature = "esp32c3",
-            feature = "esp32c6",
-            feature = "esp32h2",
-            feature = "esp32s3"
-        )))]
+    }
+}
+
+impl Printer {
+    pub fn write_bytes(bytes: &[u8]) {
         with(|| {
-            PrinterUart::write_bytes_assume_cs(bytes);
-            PrinterUart::flush();
+            PrinterImpl::write_bytes_assume_cs(bytes);
+            PrinterImpl::flush();
         })
     }
 }

--- a/esp-println/src/lib.rs
+++ b/esp-println/src/lib.rs
@@ -89,15 +89,15 @@ impl core::fmt::Write for Printer {
     }
 }
 
-#[cfg(feature="auto")]
+#[cfg(feature = "auto")]
 pub struct PrinterAuto;
 
 #[cfg(feature = "auto")]
 type PrinterImpl = PrinterAuto;
 
-#[cfg(feature="auto")]
+#[cfg(feature = "auto")]
 mod auto_printer {
-    use super::{ with, PrinterSerialJtag, PrinterUart };
+    use super::{with, PrinterSerialJtag, PrinterUart};
 
     impl super::PrinterAuto {
         fn use_jtag() -> bool {
@@ -138,7 +138,9 @@ mod auto_printer {
                 feature = "esp32p4",
                 feature = "esp32s3"
             )))]
-            { false }
+            {
+                false
+            }
         }
 
         pub fn write_bytes_assume_cs(bytes: &[u8]) {
@@ -164,7 +166,6 @@ mod auto_printer {
                 })
             }
         }
-
     }
 }
 

